docs/design/upgrade.md

```diff
@@ -0,0 +1,203 @@
+  constructor(file: string, pos: number, toRemove: string);
+}
+// Will replace text from the source code.
+class ReplaceChange implements Change {
+  constructor(file: string, pos: number, oldText: string, newText: string);
+}
+// Will output a message for the user to fulfill.
+class MessageChange implements Change {
+  constructor(text: string);
+}
+```
+
+Some higher function changes are provided to allow an easier way to create changes that are harder to make:
+
+```typescript
+// Rename a class to another name. Need to pass a symbol as input, as just
+// a string is not enough.
+declare function renameClass(project: Angular2Project,
+                             from: ts.Symbol, to: string): Observable<Change>;
+// Upgrade a package in NPM.
+declare function upgradePackage(project: Angular2Project,
+                                name: string, to: string): Observable<Change>;
+// Change a configuration in angular-cli config.
+declare function configSet(project: Angular2Project, key: string,
+                           value: any): Observable<Change>;
+// Remove a file.
+declare function deleteFile(project: Angular2Project,
+                            file: string): Observable<Change>;
+// Scaffold a new blueprints.
+declare function scaffold(project: Angular2Project, name: string,
+                          arguments: {[name: string]: string}): Observable<Change>;
+```
+
+Some higher level changes examples:
+
+```typescript
+// Add a new Import if the import isn't already in the file.
+class AddImportChange implements Change {
+  constructor(file: string, symbolName: string, importPath: string);
+}
+```
+
+# Upgrade Command
+A new command must be added to the CLI;
+
+    ng upgrade [options]
+    Upgrade the project to a newer version.
+
+    Options:
+      --package=<name>  The package to upgrade. By default, upgrade every packages
+                        that are outdated.
+      --to=<semver>     Move to a specific version of Angular. Only valid with
+                        --package and --dry-run.
+                        Default: latest.
+      --from=<semver>   Specify the version the project is currently using. Only
+                        valid with --package and --dry-run.
+                        Default: the installed version is used.
+      --commit=<bool>   Whether the tool should automatically create a commit.
+                        Default: false.
+      --message=<msg>   The commit message to use. Only valid with --commit.
+      --dry-run         Only output the commands and refactor steps that would be run,
+                        but do nothing.
+                        Default: false.
+
+## Process
+The command will execute the following tasks in order:
+
+1. If the `angular-cli` is not the latest version, return with an error message suggesting an upgrade. This step is necessary even if the version is supported as a upgrade, because updates to the upgrade steps might have occurred.
```

Comment:
I don't see the reason why this would be an error. A warning perhaps, but not an error. It forces users to *always* have the latest version if they want upgrade functionality. 

It also doesn't guarantee that the upgrade steps updates are always there either - a user can have already performed an upgrade that is currently updated. Thus we cannot operate under the assumption that upgrades are always made according to the latest definition.